### PR TITLE
[Cosmos DB] Create and manage Role Definitions and Role Assignments for enforcing data plane RBAC on Cosmos DB SQL accounts

### PR DESCRIPTION
**Pull request by nichatur**
**Description**
Adding commands for SqlRoleDefinition and SqlRoleAssignment for Cosmos DB Data Plane RBAC GA. The feature is already in  Public Preview.

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Cosmos DB

**Review by MehaKaushik — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/cosmosdb/_help.py [hunk 1]
```diff
@@ -887,3 +887,145 @@
...
+      }'
+  - name: Create a SQL role definition under an Azure Cosmos DB account using a JSON file.
+    text: az cosmosdb sql role definition create --account-name MyAccount --resource-group MyResourceGroup --body @role-definition.json
+"""
+
+helps['cosmosdb sql role definition delete'] = """
+type: command
+short-summary: Delete a SQL role definition under an Azure Cosmos DB account.
+examples:
+  - name: Create a SQL role definition under an Azure Cosmos DB account.
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/_help.py:929]:
```suggestion
  - name: Delete a SQL role definition under an Azure Cosmos DB account.
```

File: src/azure-cli/azure/cli/command_modules/cosmosdb/_validators.py [hunk 2]
```diff
@@ -61,3 +63,115 @@ def validate_virtual_network_rules(ns):
...
+
+        if 'Type' not in role_definition:
+            role_definition['Type'] = RoleDefinitionType.custom_role
+
+        role_definition['AssignableScopes'] = [_parse_resource_path(
+            scope,
+            True,
+            None,
+            get_subscription_id(cmd.cli_ctx),
+            ns.resource_group_name, ns.account_name) for scope in role_definition['AssignableScopes']]
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/_validators.py:102]: avoid positional arguments

File: src/azure-cli/azure/cli/command_modules/cosmosdb/commands.py [hunk 3]
```diff
@@ -284,3 +284,21 @@ def load_command_table(self, _):
...
+        g.command('delete', 'begin_delete_sql_role_definition', confirmation=True)
+
+    # SQL role assignment operations
+    with self.command_group('cosmosdb sql role assignment', cosmosdb_sql_sdk, client_factory=cf_sql_resources) as g:
+        g.custom_command('create', 'cli_cosmosdb_sql_role_assignment_create')
+        g.custom_command('update', 'cli_cosmosdb_sql_role_assignment_update')
+        g.custom_command('exists', 'cli_cosmosdb_sql_role_assignment_exists')
+        g.command('list', 'list_sql_role_assignments')
+        g.show_command('show', 'get_sql_role_assignment')
+        g.command('delete', 'begin_delete_sql_role_assignment', confirmation=True)
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/commands.py:304]: As they are long running operations, could we add no wait support and add wait command to query existing operation status?
https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md#register-commands (4) Supporting --no-wait, (3) Command Registration Helpers: wait_command

**Comment by Juliehzl** [comment]
Hi @nichatur, I want to check with you that if these commands exist in a cli extension. If yes, after PR merge, please remember to remove same commands in cli extension.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/commands.py (hunk 3 above)
Comment (nichatur) [reply src/azure-cli/azure/cli/command_modules/cosmosdb/commands.py]: Thanks, added the command

**Comment by nichatur** [comment]
> Hi @nichatur, I want to check with you that if these commands exist in a cli extension. If yes, after PR merge, please remember to remove same commands in cli extension.
Hi @Juliehzl , yes these commands are in the cosmosdb-preview cli extension. I will work on a PR to remove them.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py [hunk 4]
```diff
@@ -1612,3 +1621,144 @@ def cli_cosmosdb_collection_update(client,
...
+                                            account_name,
+                                            role_definition_body):
+    '''Creates an Azure Cosmos DB SQL Role Definition '''
+    role_definition_create_resource = SqlRoleDefinitionCreateUpdateParameters(
+        role_name=role_definition_body['RoleName'],
+        type=role_definition_body['Type'],
+        assignable_scopes=role_definition_body['AssignableScopes'],
+        permissions=role_definition_body['Permissions'])
+
+    return client.begin_create_update_sql_role_definition(role_definition_body['Id'], resource_group_name, account_name, role_definition_create_resource)
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py:1637]: with support_no_wait=True, you also need to use sdk_no_wait for all your long running operation functions. e.g. 
https://github.com/Azure/azure-cli/blob/4a5d74215d5d1566f839c2d11fd360d300f56b62/src/azure-cli/azure/cli/command_modules/network/custom.py#L260

Comment (Juliehzl) [reply src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py:1637]: Please add a test for command with `--no-wait` to make sure it work as expected.

File: src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py [hunk 5]
```diff
@@ -1612,3 +1615,141 @@ def cli_cosmosdb_collection_update(client,
...
+    if role_definition_body['Permissions'] is not None:
+        role_definition.permissions = role_definition_body['Permissions']
+
+    role_definition_update_resource = SqlRoleDefinitionCreateUpdateParameters(
+        role_name=role_definition.role_name,
+        type=role_definition_body['Type'],
+        assignable_scopes=role_definition.assignable_scopes,
+        permissions=role_definition.permissions)
+
+    return client.begin_create_update_sql_role_definition(role_definition_body['Id'], resource_group_name, account_name, role_definition_update_resource)
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/cosmosdb/custom.py:1657]: miss sdk_no_wait here?

**Review by Juliehzl — APPROVED** [review]

**Comment by Juliehzl** [comment]
One more suggestion. As it is the first time that you add it in main repo and the feature is in public preview not GA, could we add preview tag for your feature now?

**Comment by Juliehzl** [comment]
> One more suggestion. As it is the first time that you add it in main repo and the feature is in public preview not GA, could we add preview tag for your feature now?
> https://github.com/Azure/azure-cli/blob/fb60a21d0abaa533e36fbec0974b59bae4f37d75/src/azure-cli/azure/cli/command_modules/storage/commands.py#L153-L154

Discussed online. It will go to GA next week. Approved with no Preview tag.

**Pull request merged**